keps/sig-node/20200303-shared-pod-limits.md

```diff
@@ -0,0 +1,326 @@
+*  Providing a general-purpose interface to the full range of possible resource management provided by the Linux cgroup hierarchy.
+
+## Proposal
+
+The Pod QoS enhancement already implemented in Kubernetes manages resources as a hieararchy of cgroups in the following way:
+
+<pre>
+  QoS CGroup (one of guaranteed, burstable, or besteffort)
+   |
+   \ pod 
+      |
+      +-- container0 (pause container)
+      |
+      +-- container1 (first container)
+      |
+   = ... =
+      |
+      +-- containerN (N-th container)
+</pre>
+
+Each container level cgroup is limited based on the information provided in the `Container` specification. Currently, if the pod belongs to 
+the `Burstable` QoS level, and **all** of the containers specify limits for each resource, then these limits are summed, and that limit is
+assigned to the pod level cgroup.
+
+The implications of the current `Burstable` definition is that if any containers belonging to the Pod don't define a limit, then those containers
+are effectively not limited by the Linux kernel. Conversely, if all of the containers provided a limit, then even though the pod level cgroup is 
+configured with the sum of those limits, there is no significance to this since no container can ever use more of the resource than what is defined
+in the container level cgroup.
+
+The proposal in this KEP is to allow users to opt-in to a slightly modified definition of the `Burstable` QoS level. In this modified definition,
+the sum of all defined container limits for each resource are always assigned to the pod-level cgroup. This is done by adding an attribute called 
+`shareBurstableLimits` on the Pod specification level. If this attribute is set to false (the default) the feature is disabled, and if it is set to 
+true then the feature takes effect.
+
+For example, consider the following Pod:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    run: nginx
+  name: nginx
+spec:
+  containers:
+  - image: shell
+    name: debian:buster
+  - image: proxy
+    name: envoy:latest
+  - name: nginx
+    image: nginx:latest
+    command: [ "/usr/bin/tail", "-f", "/dev/null"] 
+    resources:
+      requests:
+        memory: 128M
+        cpu: "0.5"
+      limits:
+        memory: 256M
+        cpu: "1"
+```
+
+This pod belongs to the `Burstable` QoS level because it defines multiple containers, of which only one has resource limits, and there is a 
+difference between the `Request` and the `Limit` for both these resources. Effectively, the containers which don't define any limits are
+not constrained in any way. They are provided the same level of resources as a container in a `BestEffort` pod would, except that they belong to a pod with a higher priority, so they are more proofed from causing evictions.
+
+The cgroup hierarchy for each of these resources (memory and cpu) would be this:
+
+<pre>
+  QoS CGroup (one of guaranteed, burstable, or besteffort)
+   |
+   \ pod (memory: unlimited, cpu: unlimited)
```

Comment:
for clarity, cpu is not unlimited... it is setting shares, just quota is unbounded.

Reply:
Correct, I will fix the KEP.